scripts/qa_crowbarsetup.sh

```diff
@@ -3860,6 +3908,98 @@ function oncontroller_heat_image_setup()
     fi
 }
 
+function oncompute_parse_vm_xml
+{
+    local instance_name=$1
+    echo $instance_name
+    virsh list --all | grep $instance_name
+    [ $? != 0 ] && complain 53 "failed to find pci-passthru-instance"
+
+    # verify node xml to see if it has pci passthru section which has the
+    # specific address passing from compute host
+    virsh dumpxml $instance_name | grep -E "<hostdev" -A 7 | grep "0x1d"
+    [ $? != 0 ] && complain 63 "failed to find pci-passthru block in node xml"
+
+    return 0
+}
+
+function oncontroller_retrieve_image
+{
+    local image_name=$1
+    local image_url=$2
+
+    if [ -n "${localreposdir_target}" ] ; then
+        local_image_path=$localreposdir_target/images/$arch/openstack/$image_name
+    fi
+
+    if [ -n "${local_image_path}" -a -f "${local_image_path}" ] ; then
+        image_name=$local_image_path
+    else
+        local ret=$(wget -N --progress=dot:mega "$image_url" 2>&1 >/dev/null)
+        if [[ $ret =~ "200 OK" ]]; then
+            echo $ret
+        elif [[ $ret =~ "Not Found" ]]; then
+            complain 73 "$image_name image not found: $ret"
+        else
+            complain 74 "failed to retrieve $image_name image: $ret"
+        fi
+    fi
+}
+
+function oncontroller_upload_image_to_glance
+{
+    local image_name=$1
+    local image_path=$2
+    shift ; shift
+    local image_params=$@
+
+    if [ -z "$image_params" ]; then
+        image_params="--disk-format qcow2 --property hypervisor_type=kvm"
+    fi
+
+    . .openrc
+
+    # using list subcommand because show requires an ID
+    if ! openstack image list --format value -c Name | grep -q "$image_name"; then
+        openstack image create --file $image_path \
+            $image_params --container-format bare --public \
+            $image_name
+    fi
+}
+
+function oncontroller_setup_pci_passthru_vm
```

Comment:
I was trying to reuse the `testvm` generated from `testsetup` step for past few days. However, I run into situation that the `testvm` won't boot up properly if the device passing from compute is attached (the specific flavor which has pci passthru attribute set). The workaround will be tweaking it to unplug/replug the device during post boot up phase. So I go back to my original formula instead of touching too many parts in `oncontroller_testsetup`.